### PR TITLE
Move to src-layout.

### PR DESCRIPTION
**Pull request by apollo13**
The intention of this is better isolation. `trio` is no longer accidentally on the (python) path and as such requires an explicit installation for usage, this helps uncover issues in packaging data etc early on (see
https://packaging.python.org/en/latest/discussions/src-layout-vs-flat-layout/).

This serves as the basis for switching to another build system as well; refs #2790.

### CODE REVIEW

**Comment by altendky** [comment]
In CI this is already addressed by changing into a subdirectory when testing.  I personally prefer `src/`, but not doing that has been an intentional decision thus far.  Though, I forget where I saw this discussed.  `:[`  I'm just suggesting that the discussion is had since I don't see a reference from here to it..

**Comment by altendky** [comment]
https://github.com/python-trio/trio/blob/c16003f08285d0b5b3b5b16d5bdc4f00b1e3cd52/ci.sh#L106-L110

**Comment by apollo13** [comment]
@altendky I see, I will still finish this for now and put it up for discussion, because while CI might address it, it can still be pretty much an annoyance. Let's see what the rest says :)

**Comment by jakkdl** [comment]
I think #274 is the only place where it's been discussed

**Comment by CoolCat467** [comment]
I definitely agree, a `src/` based layout would be much nicer, +1

**Comment by jakkdl** [comment]
@njsmith do you want to weigh in on the matter?

**Comment by jakkdl** [comment]
@A5rocks any opinion on the matter?

**Comment by A5rocks** [comment]
I'm neutral on this change cause while it does align with what most people do, so would e.g. not shipping tests... so I don't think that's a very convincing argument.

And while it would take a bit of magic away from the tests it would also mean having an extra "src" folder in the path for every file in trio. (Then of course, I already have to type out trio/trio for the folder...)

So yeah arguments for and against but none of them feel strong, so neutral.

**Comment by jakkdl** [comment]
if this makes switching packaging system easier, I think we can just go ahead with it. The PRs been open a month and the only issue that has been brought up are minor workflow ones.

**Comment by CoolCat467** [comment]
@apollo13 I would be glad to finish working on getting this merged this if you would like

**Comment by apollo13** [comment]
@CoolCat467 That would be great, I am currently blocked by personal things. My goal with the PR was to be as minimal as possible. Ie moving `_tests` (if wanted at all) into another top level package would be something done after this PR (same for switching to whatever other tooling). I can give you access to my repo if you want to work on that branch directly or we close this PR and you open your own.

**Comment by apollo13** [comment]
> if this makes switching packaging system easier, I think we can just go ahead with it. The PRs been open a month and the only issue that has been brought up are minor workflow ones.

I guess it doesn't make switching much easier (if at all), tooling can usually handle both.

**Comment by altendky** [comment]
> Ie moving `_tests` (if wanted at all) into another top level package

I don't want that.  `:]`  I like the stuff I import to have been installed so it's not all fiddly about how you launch Python or what your cwd is.  Also, why not be able to run tests on deployed packages?  Also also, fewer directories to both including and excluding from all the myriad of tools and packaging definitions etc.

**Comment by CoolCat467** [comment]
> @CoolCat467 That would be great, I am currently blocked by personal things. My goal with the PR was to be as minimal as possible. Ie moving `_tests` (if wanted at all) into another top level package would be something done after this PR (same for switching to whatever other tooling). I can give you access to my repo if you want to work on that branch directly or we close this PR and you open your own.

If I could have access to the repo that would be great, I would prefer not creating another PR if it could be avoided.

**Comment by apollo13** [comment]
Sent an invite

**Comment by A5rocks** [comment]
fyi (cause review was requested) i want to merge this in after the next trio release, in order to minimize possible disruption.... unless someone wants this urgently

**Comment by A5rocks** [comment]
OK release is out. Still ambivalent but if people think this would better and there isn't dissent I guess this should happen? Even if we don't end up using a pep 612 backend that requires the `src/` dir, it still might be better to stick to the beaten path here.

(Obviously conflicts need to be fixed)

**Review by CoolCat467 — APPROVED** [review]
Slightly maybe in warning of "review your own PRs" since I've taken over continued work on this, but all the important changes were already made before that point and the tests are passing.

**Review by jakkdl — APPROVED** [review]
we might want to wait a day to see if anybody else pipes up about trio 0.23 being broken and needing a hotfix patch, but otherwise I don't see any problems with merging this.

File: .coveragerc
Comment (A5rocks) [inline .coveragerc:1]: We'll need to move this to `pyproject.toml`

File: setup.py
Comment (A5rocks) [inline setup.py:1]: We dropped this so you might need to wrangle the new config in `pyproject.toml`

**Review by A5rocks — COMMENTED** [review]
Conflicts again :(

But mainly just that new exciting `pyproject.toml` changes mean you have to migrate some of the changes here.

File: .coveragerc
Comment (A5rocks) [reply .coveragerc:1]: Oops, you're still creating `.coveragerc`

**Comment by CoolCat467** [comment]
@A5rocks How do we fix this error?
```console
  + coverage report -m --rcfile ../pyproject.toml
  No data to report.
  Error: Process completed with exit code 1.
```

**Comment by A5rocks** [comment]
Make sure that coverage is outputting its data files in the cwd, maybe? Not sure tbh

**Comment by CoolCat467** [comment]
@A5rocks I fixed the coverage report and resolved conflicts

File: pyproject.toml
Comment (A5rocks) [inline pyproject.toml:236]: I presume these /src/ changes are for source code? But don't we run coverage against trio as a package? I hope we're not shipping /src/ in the wheel...

**Review by A5rocks — COMMENTED** [review]
GitHub says I have a pending review so let's see what it is...

File: pyproject.toml
Comment (CoolCat467) [reply pyproject.toml]: Should be fixed in 3b582ca

**Comment by jakkdl** [comment]
let's not delay this any further and fix any other potential errors as they crop up. :rocket:

**Pull request merged**
